New integration: TED (replaced TED5000, adds support for TED6000)

rianadon opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

The `ted5000` integration will be replaced with the The Energy Detective (`ted`) integration. The ted5000 integration will need to be removed from `configuration.yaml`, and the `ted` integration must be configured via the UI.


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Replaces the TED5000 integration with a new TED integration supporting both the TED5000 and TED6000 energy monitors.

This is my first integration, so I based the code off both the TED5000 integration and the Enphase Envoy integration.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

Brands pull request: https://github.com/home-assistant/brands/pull/2800
Documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19234

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rianadon committed:
Adapt ted5000 to ted6000

rianadon committed:
Add config flow and coordinator similar to enphase integration

rianadon committed:
Clean up config flow

rianadon committed:
Add test

gtdiehl commented:
@rianadon Have you created PRs for the documentation and brands, in these repositories [documentation](https://github.com/home-assistant/home-assistant.io) and [brands](https://github.com/home-assistant/brands)? This is so the instructions will appear on the website with the logo and the logo and icon will appear in Home Assistant.

rianadon committed:
Configure with CONF_HOST, remove envoy references

rianadon commented:
Brands pull request: https://github.com/home-assistant/brands/pull/2800
Documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19234

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

rianadon committed:
unit_of_measurement -> native_unit_of_measurement

ludeeus commented on homeassistant/components/ted6000/config_flow.py:
```diff
@@ -0,0 +1,115 @@
+
+            try:
+                async with get_async_client(self.hass):
+                    settings = httpx.get(
+                        f"http://{user_input[CONF_HOST]}/api/SystemSettings.xml"
+                    )
+                doc = xmltodict.parse(settings.text)["SystemSettings"]
```
Communication logic needs to be extracted to a standalone library hosted on PyPI.

ludeeus commented on homeassistant/components/ted6000/manifest.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "domain": "ted6000",
```
We should not add a new domain for each model.
We should have `ted` as the domain that works with both models.

ludeeus commented on homeassistant/components/ted6000/const.py:
```diff
@@ -0,0 +1,38 @@
+COORDINATOR = "coordinator"
+NAME = "name"
+
+SENSORS = (
```
This is specific to the sensor platform, move it to the sensor.py file.

rianadon committed:
Merge ted5000 & ted6000 + host communication logic on pypi

rianadon committed:
Mock tedpy in tests and upgrade

rianadon committed:
Increase code coverage

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

rianadon committed:
Increase test coverage even more

rianadon committed:
Remove reauth code

rianadon committed:
Remove placeholders since it looks like there's no paths for them

rianadon committed:
Add MTU & Spyder data reporting

rianadon committed:
Add translations

rianadon committed:
Decrease async timeouts

rianadon committed:
Don't publish daily consumption for spyders, state class total

rianadon committed:
Use model instead of ID to name integration

rianadon committed:
Use native_value

rianadon committed:
Allow spyder devices to be configured via options

rianadon committed:
Add test for options flow

rianadon committed:
Upgrade tedpy

rianadon committed:
Remove non-English translations

I think I read somewhere these are automatically generated by the
translation tool so should not be comitted for now?

rianadon committed:
Upgrade tedpy and support mtu historical data

rianadon commented:
With this last commit I am finished writing features for the integration. All that's left is getting it ready for merge!

rianadon commented:
Hi @ludeeus / @gtdiehl. No rush, but the PR is ready for review if you have the chance.

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

rianadon commented:
Sorry I had to take a moment to update the PR to the latest dev changes and accidently merged in master which screwed up reviewers & commit history. I'm hoping everything is back to normal now.

thecode commented:
> Sorry I had to take a moment to update the PR to the latest dev changes and accidently merged in master which screwed up reviewers & commit history. I'm hoping everything is back to normal now.

You need to rebase your branch to dev

rianadon committed:
Remove sensor.py from ted5000

rianadon commented:
> > Sorry I had to take a moment to update the PR to the latest dev changes and accidently merged in master which screwed up reviewers & commit history. I'm hoping everything is back to normal now.
> 
> You need to rebase your branch to dev

Thanks! I reverted back to the commit before the merge, re-merged, then force-pushed. It looks like from the file changes everything is alright and normal again.

Ramias1 commented:
Just checking in on this -- is work still being done to get this in to HA core?  Would love to be able to use my Ted6000 with HA on docker.  Thanks

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

Ramias1 commented:
Checking to see if this is still being worked on.  Would love to see it.  

If the code exists can anybody please provide guidance on how to get it to work with Home Assistant core?

realumhelp commented:
> Checking to see if this is still being worked on. Would love to see it.
> 
> If the code exists can anybody please provide guidance on how to get it to work with Home Assistant core?

[rianadon](https://github.com/rianadon) has been leading the charge. The backend project is here, give it a shot see if it works: 
 * https://github.com/rianadon/the-energy-detective-py

@rianadon has written the HA integration pieces as well and going back and forth with code reviews/change requests. So some progress overall. I don't know enough about HA Integration code to be of much help at this point, although I'd really like too.

Ramias1 commented:
> > Checking to see if this is still being worked on. Would love to see it.
> > If the code exists can anybody please provide guidance on how to get it to work with Home Assistant core?
> 
> [rianadon](https://github.com/rianadon) has been leading the charge. The backend project is here, give it a shot see if it works:
> 
> * https://github.com/rianadon/the-energy-detective-py
> 
> @rianadon has written the HA integration pieces as well and going back and forth with code reviews/change requests. So some progress overall. I don't know enough about HA Integration code to be of much help at this point, although I'd really like too.

Thanks.  I have copied the ted folder from @rianadon's repo and added it to my custom_components folder.  I've checked permissions on it but when I try to add a ted sensor in configuration.yaml and then check the yaml configuration it fails with "Platform error sensor.ted - Integration 'ted' not found.".  I've restarted a few times.  Any idea what I'm doing wrong?   If I add the ted5000 sensor it just pulls from home assistant's existing ted5000 code and fails since I have a ted 6000.

Ramias1 commented:
Adding this to the energy dashboard.  I'm adding my daily net energy sensor and the console shows:

```
The following entities have state class 'measurement' but 'last_reset' is missing:
sensor.160345_today_s_net_energy
```

Not sure where in the code to add "last_reset" but may be good to do this for the daily and monthly counters.

Ramias1 commented:
Should also note:  getting this in my logs.

`Logger: homeassistant.helpers.frame
Source: helpers/frame.py:77 
First occurred: 9:45:24 AM (1 occurrences) 
Last logged: 9:45:24 AM

Detected integration that uses deprecated `async_get_registry` to access entity registry, use async_get instead. Please report issue to the custom integration author for ted using this method at custom_components/ted/sensor.py, line 69: entity_registry = await hass.helpers.entity_registry.async_get_registry()`

rianadon commented:
Many component-facing APIs in Home Assistant have changed since I wrote the PR. It's been difficult keeping up with the changes, especially since I have no idea if/when this will be merged.

The `last_reset` is probably a property of the entity classes that needs to be defined. It seems like for the daily counters, this could be set to midnight.

If there's enough interest and someone wants to help maintain the repository, I'll consider making this a custom component that can be installed through HACS.

Ramias1 commented:
> Many component-facing APIs in Home Assistant have changed since I wrote the PR. It's been difficult keeping up with the changes, especially since I have no idea if/when this will be merged.
> 
> The `last_reset` is probably a property of the entity classes that needs to be defined. It seems like for the daily counters, this could be set to midnight.
> 
> If there's enough interest and someone wants to help maintain the repository, I'll consider making this a custom component that can be installed through HACS.

I'm probably not skilled enough with Python or Home Assistant to be of much use but I'd be happy to help test it and would love it if this were available in HACS.  Though there may only be a few of us using it :).  Thanks again for the great work.

Ramias1 committed:
Update sensor.py; minor changes

Changed line 69 to use async_get to suppress log error.  

Changed line 38 state_class to TOTAL_INCREASING so energy monitor will work with energy dashboard.

Ramias1 committed:
Merge pull request #1 from Ramias1/Ramias1-patch-1

Update sensor.py; minor changes

natethelen commented:
I was about to start work on this very thing (even asked community on discord if I should create a new one or modify the ted5000 one).  I am happy to put some time into this to help get it over the finish line.  @rianadon do you have a feeling of what is holding it up?  Is it the fact it is a breaking change?

natethelen commented:
I will see if I can get these 3 failing checks fixed to start

rianadon commented:
It's been a while since I've merged in the latest Home Assistant development branch into my work, so the conflicts and failed tests are likely due to the changing Home Assistant APIs.

I've also done a bit of thinking since and I believe this PR might be more palatable if we created the new integration then marked the old one as deprecated (maybe using the new repairs panel?) and encourage everyone using the `ted5000` integration to switch to the `ted` one.

I haven't had a chance to look into the code needed to do this, but I can try my hand at fixing up the existing code.

natethelen commented:
Yes, I believe that would be much safer.  My experience using HA is that I would be upset if my installation broke by doing an update.

In order to help out on this I think you are going to have to give me write access to the ted6000 branch.  I am happy to spend time on it

Ramias1 commented:
Great to see others looking into this.  I hacked the sensor.py a bit here: https://github.com/Ramias1/core/tree/ted6000/homeassistant/components/ted and have this working and reporting to the Energy Dashboard for my two Spyders.  I tried to do a PR to Riandon's branch but got overwhelmed with the HA contributor language; feel free to look at the link above and compare to Riandon's branch.  Was only a few lines of changes.

rianadon commented:
Thanks for making those changes @Ramias1! They all make sense to me. I'll see if I can try to pull them in to my version.

Ramias1 commented:
just signed the CLA.  hopefully that moves this forward.

rianadon commented:
Since the new approach of keeping the TED5000 integration is significantly different than what I have here, I made a new pull request at #82959. I'm going to close this one.
